File: android/rctmgl/src/main/java/com/mapbox/rctmgl/components/camera/CameraUpdateQueue.java
package com.mapbox.rctmgl.components.camera;

import com.mapbox.mapboxsdk.maps.MapboxMap;

import java.util.Iterator;
import java.util.LinkedList;
import java.util.Queue;

/**
 * Created by nickitaliano on 9/5/17.
 */

public class CameraUpdateQueue {
    private Queue<CameraStop> mQueue;
    private OnCompleteAllListener mCompleteListener;

    public interface OnCompleteAllListener {
        void onCompleteAll();
    }

    public CameraUpdateQueue() {
        mQueue = new LinkedList<>();
    }

    public void offer(CameraStop item) {
        mQueue.offer(item);
    }

    public int size() {
        return mQueue.size();
    }

    public boolean isEmpty() {
        return mQueue.isEmpty();
    }

    public void flush() {
        while (!mQueue.isEmpty()) {
            mQueue.remove();
        }
    }

    public void setOnCompleteAllListener(OnCompleteAllListener listener) {
        mCompleteListener = listener;
    }

    public void execute(final MapboxMap map) {
        if (mQueue.isEmpty()) {
            if (mCompleteListener != null) {
                mCompleteListener.onCompleteAll();
            }
            return;
        }

        final CameraStop stop = mQueue.poll();
        if (stop == null) {
            return;
        }

        final CameraUpdateItem item = stop.toCameraUpdate();
        item.execute(map, new CameraUpdateItem.OnCameraCompleteListener() {
            @Override
            public void onComplete() {
                execute(map);
            }
        });
    }
}

File: android/src/main/java-v10/com/mapbox/rctmgl/components/camera/CameraUpdateQueue.kt
package com.mapbox.rctmgl.components.camera

import com.mapbox.rctmgl.components.camera.CameraStop.Companion.fromReadableMap
import com.mapbox.maps.MapboxMap
import com.mapbox.maps.CameraOptions
import com.mapbox.maps.plugin.animation.CameraAnimationsPlugin
import android.view.animation.LinearInterpolator
import android.view.animation.AccelerateDecelerateInterpolator
import com.mapbox.rctmgl.components.camera.CameraStop
import com.mapbox.rctmgl.components.camera.CameraUpdateQueue.OnCompleteAllListener
import com.mapbox.rctmgl.components.mapview.RCTMGLMapView
import com.mapbox.rctmgl.components.camera.CameraUpdateItem
import com.facebook.react.bridge.ReactApplicationContext
import com.mapbox.rctmgl.components.AbstractEventEmitter
import com.mapbox.rctmgl.components.camera.RCTMGLCamera
import com.mapbox.rctmgl.components.camera.RCTMGLCameraManager
import com.facebook.react.uimanager.ThemedReactContext
import com.facebook.react.uimanager.annotations.ReactProp
import com.facebook.react.bridge.ReadableMap
import java.util.*

class CameraUpdateQueue {
    private var mQueue: Queue<CameraStop>
    private var mCompleteListener: OnCompleteAllListener? = null

    interface OnCompleteAllListener {
        fun onCompleteAll()
    }

    init {
        mQueue = LinkedList()
    }

    fun offer(item: CameraStop) {
        mQueue.offer(item)
    }

    fun size(): Int {
        return mQueue.size
    }

    val isEmpty: Boolean
        get() = mQueue.isEmpty()

    fun flush() {
        while (mQueue.size > 0) {
            mQueue.remove()
        }
        mQueue = LinkedList()
    }

    fun setOnCompleteAllListener(listener: OnCompleteAllListener?) {
        mCompleteListener = listener
    }

    fun execute(map: RCTMGLMapView?) {
        if (mQueue.isEmpty()) {
            mCompleteListener?.let { it.onCompleteAll() }
            return
        }
        val stop = mQueue.poll() ?: return
        val item = stop.toCameraUpdate(map!!)
        item.run()
        execute(map)
    }
}